backend/controllers/bulkAppointmentsUpload.js

```diff
@@ -0,0 +1,86 @@
+const bulkAppointmentsUploadFactory = (csvParserService, offenderLoader, logError) => {
+  const index = async (req, res) => {
+    try {
+      if (!req.session.data) {
+        res.render('error.njk', {
+          url: '/need-to-upload-file',
+        })
+        return
+      }
+
+      res.render('bulkUploadFile.njk', {
+        title: 'Upload a CSV File',
+        errors: req.flash('errors'),
+        appointmentDetails: req.session.data,
+      })
+    } catch (error) {
+      logError(req.originalUrl, error, 'Sorry, the service is unavailable')
+      res.render('error.njk', {
+        url: '/need-to-upload-file',
+      })
+    }
+  }
+
+  const getNonExistingOffenderNumbers = (uploadedList, prisonersList) => {
+    const matchingPrisonerNumbers = []
+    uploadedList.forEach(uploadedOffenderNo =>
+      prisonersList.forEach(prisoner => {
+        if (uploadedOffenderNo === prisoner.offenderNo) {
+          matchingPrisonerNumbers.push(uploadedOffenderNo)
+        }
+      })
+    )
+
+    return uploadedList.filter(x => matchingPrisonerNumbers.indexOf(x) < 0)
+  }
+
+  const postAndParseCsvOffenderList = async (req, res) => {
+    const { file } = req.files
+    const { location } = req.session.data
+
+    csvParserService
+      .loadAndParseCsvFile(file)
+      .then(async fileContent => {
+        const prisonersDetails = await offenderLoader.loadFromCsvContent(res.locals, fileContent, location)
+
+        const removeDuplicates = array => [...new Set(array)]
+        const offendersFromCsv = removeDuplicates(fileContent.map(row => row[0]))
+
+        const prisonerList = prisonersDetails.map(prisoner => ({
+          bookingId: prisoner.bookingNo,
+          offenderNo: prisoner.offenderNo,
+          firstName: prisoner.firstName,
+          lastName: prisoner.lastName,
+        }))
+
+        const offenderNosNotFound = getNonExistingOffenderNumbers(offendersFromCsv, prisonerList)
+
+        if (offenderNosNotFound.length === fileContent.length) {
+          throw new Error('Select a file with prison numbers in the correct format - for example, A1234BC')
+        }
+
+        if (prisonerList && prisonerList.length) {
+          // eslint-disable-next-line no-param-reassign
+          req.session.data.prisonersListed = prisonerList
+        }
+
+        if (offenderNosNotFound && offenderNosNotFound.length) {
+          // eslint-disable-next-line no-param-reassign
+          req.session.data.prisonersNotFound = offenderNosNotFound
+        }
+
+        return res.redirect('/bulk-appointments/confirm-appointments')
```

Comment:
There is a slight issue here whereby both of the above checks fail and we still redirect.